py/qstr.h

```diff
@@ -47,12 +47,30 @@ enum {
 
 typedef size_t qstr;
 
+#if MICROPY_QSTR_BYTES_IN_HASH == 1
+typedef uint8_t qstr_hash_t;
+#elif MICROPY_QSTR_BYTES_IN_HASH == 2
+typedef uint16_t qstr_hash_t;
+#else
+#error unimplemented qstr hash decoding
+#endif
+
+#if MICROPY_QSTR_BYTES_IN_LEN == 1
+typedef uint8_t qstr_len_t;
+#elif MICROPY_QSTR_BYTES_IN_LEN == 2
+typedef uint16_t qstr_len_t;
+#else
+#error unimplemented qstr length decoding
+#endif
+
 typedef struct _qstr_pool_t {
     struct _qstr_pool_t *prev;
     size_t total_prev_len;
     size_t alloc;
     size_t len;
-    const byte *qstrs[];
+    qstr_hash_t *hashes;
+    qstr_len_t *lengths;
+    const char *qstrs[];
```

Comment:
Looks like this means that all qstr pools in RAM take up an extra 2x words.  But it still fits within 2x GC blocks (4 words per block) so shouldn't have any effect on RAM usage (would only increase size of pools in flash).